google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageImpl.java

```diff
@@ -682,9 +750,78 @@ private Blob internalCreate(
 
   /** Bind some decoders for our "Syntax" classes to this instance of GrpcStorageImpl */
   private final class SyntaxDecoders {
+
     final Decoder<Object, Blob> blob =
         o -> codecs.blobInfo().decode(o).asBlob(GrpcStorageImpl.this);
     final Decoder<com.google.storage.v2.Bucket, Bucket> bucket =
         b -> codecs.bucketInfo().decode(b).asBucket(GrpcStorageImpl.this);
   }
+
+  private static class TransformPageDecorator<
+          RequestT,
+          ResponseT,
+          ResourceT,
+          PageT extends AbstractPage<RequestT, ResponseT, ResourceT, PageT>,
+          ModelT>
+      implements Page<ModelT> {
+
+    private final PageT page;
+    private final Function<ResourceT, ModelT> translator;
+
+    public TransformPageDecorator(PageT page, Function<ResourceT, ModelT> translator) {
+      this.page = page;
+      this.translator = translator;
+    }
+
+    @Override
+    public boolean hasNextPage() {
+      return page.hasNextPage();
+    }
+
+    @Override
+    public String getNextPageToken() {
+      return page.getNextPageToken();
+    }
+
+    @Override
+    public Page<ModelT> getNextPage() {
+      return new TransformPageDecorator<>(page.getNextPage(), translator);
+    }
+
+    @Override
+    public Iterable<ModelT> iterateAll() {
+      return () -> {
+        final Iterator<ResourceT> iter = page.iterateAll().iterator();
+        return new ResourceIterator(iter);
+      };
+    }
+
+    @Override
+    public Iterable<ModelT> getValues() {
+      return () -> {
+        final Iterator<ResourceT> inter = page.getValues().iterator();
+        return new ResourceIterator(inter);
+      };
+    }
+
+    private class ResourceIterator implements Iterator<ModelT> {
```

Comment:
Nit: `s/ResourceIterator/TransformingIterator/g`

The interesting work this iterator is doing is the transformation, not really anything Resource related.